source/xep/mam/MamManager.cpp

```diff
@@ -107,135 +115,140 @@ void MamManager::requestArchiveForJid(const QString& jid, const QString &last)
 
         Swift::JID sJid(jid.toStdString());
 
-        //qDebug() << xw.getXmlResult();
-
-        client_->getIQRouter()->sendIQ(Swift::IQ::createRequest(Swift::IQ::Set, sJid, msgId,
-                                                                std::make_shared<Swift::RawXMLPayload>(xw.getXmlResult().toStdString())
-                                                                ));
+        // Calling getXmlResult flushes content and subsequent calls are empty! 
+        // qDebug() << xw.getXmlResult(); 
 
+        MamRequest::ref mamRequest = MamRequest::create(Swift::IQ::Set, sJid, xw.getXmlResult().toStdString(), client_->getIQRouter());
+        mamRequest->onResponse.connect(boost::bind(&MamManager::processFinIq, this, _1, _2, _3));
+        mamRequest->send();
     }
 }
 
-// FIXME rewrite me!
-// this fails sometimes :-(.
-// use custom payload parser to filter out mam messages!
-
-void MamManager::handleDataReceived(Swift::SafeByteArray data)
+void MamManager::handleMessageReceived(Swift::Message::ref message)
 {
-    std::string nodeData = Swift::safeByteArrayToString(data);
-    QString qData = QString::fromStdString(nodeData);
+    auto mamResult = message->getPayload<Swift::MAMResult>();
```

Comment:
A crazy idea. mamResult is from type Swift::Forwarded. What if Shmoose get rid of MamManager completely? And the small adjustment will be done in MessageHandler::handleMessageReceived? There is already an implementation which handles the Swift::Forwarded messages. Maybe this fits.

Reply:
Let me check but yes why not.

Reply:
Hi, I did the change to manage Mam messages in MessageHandler::handleMessageReceived. I agree this is better for maintenance but likely also for performance. I checked in but I need to do more tests.

Reply:
Agree. Just had a look at the source. This looks much cleaner now. I will also do some functional testing the next days.